### PR TITLE
Fix env propagation

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes https://github.com/openshift/jenkins-client-plugin/issues/180
Fixes https://github.com/openshift/jenkins-client-plugin/issues/177

@openshift/sig-developer-experience @vfreex ptal

### CODE REVIEW

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java [hunk 1]
```diff
@@ -45,7 +48,38 @@ public ClientCommandBuilder(String server, String project, boolean skipTLSVerify
...
+                // a) has neither ' or "", or
+                // b) has ", but doesn not start with it, and does not start with '
+                // then wrapper with '
+                if ((!s.contains("'") && !s.contains("\"")) ||
+                   (!s.startsWith("\"") && !s.startsWith("'") && s.contains("\"")))
+                    s = "'" + s + "'";
+                // if the arg that comes in either 
+                // c) has ', but does not start with it, and does not start with "
+                // then wrapper with "
+                if (!s.startsWith("\"") && !s.startsWith("'") && s.contains(("'")))
```
Comment (bparees) [inline src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java]: you are doing this check against the string you just (potentially) added quotes to.  intentional?

Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java]: ahh ... thx, will address

**Comment by gabemontero** [comment]
update for first comment pushed @bparees

**Comment by bparees** [comment]
lgtm.

**Comment by gabemontero** [comment]
thx @bparees 

and yep, let us let @vfreex chime in before actually merging, 

but based on my testing, his examples like

`openshift.selector('po', '|| ping google.com -c 5 > out.txt ; echo {} | tee --').object()`

should be ok.  But certainly if he wants to check out the branch and give it a go himself, let's give him some time to do that, though we do have a fair number of upstream users impacted, so let's not wait super long.

Also, once we have a new version of the plugin available for the PR tester, I do have a simple reproducer or two we can add to the test pipeline.

**Comment by bparees** [comment]
> Also, once we have a new version of the plugin available for the PR tester, I do have a simple reproducer or two we can add to the test pipeline.

great, tests was gong to be my next question :)

**Comment by gabemontero** [comment]
ahhh ... I forgot to validate the rsh/exec path (and @vfreex mentioned that earlier today)

OK, I'll circle back to that on Wednesday.  Might need to fall back on verb based special casing I was hoping to avoid, but we'll see ...

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java [hunk 2]
```diff
@@ -96,7 +102,44 @@ public void checkRoles(RoleChecker checker) throws SecurityException {
...
+                if (path == null || path.length() == 0) {
+                    LOGGER.warning("PATH not properly set prior to invocation of 'oc'");
+                } else {
+                    String[] dirs = path.split(File.pathSeparator);
+                    boolean override = false;
+                    for (String dir : dirs) {
+                        File f = new File(dir);
+                        File[] fileList = f.listFiles();
+                        if (fileList != null)
+                            for (File file : fileList) {
```
Comment (vfreex) [inline src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java]: Is there any particular reason to iterate over all files in the inspected directory? Can we use some condition like `if (new File(path, prog.toString()).canExecute())`?

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java [hunk 3]
```diff
@@ -45,7 +48,39 @@ public ClientCommandBuilder(String server, String project, boolean skipTLSVerify
             return n;
         }
         for (Object o : l) {
-            n.add(o.toString());
+            String s = o.toString();
+            if (s.contains(" ")) {
+                // the Jenkins QuotedStringTokenizer is not quite 
+                // copacetic with template parameters defined in 
+                // groovy variables (specifically strings where you have spaces but don't have 
+                // to encapsulate with " or ').
```
Comment (vfreex) [inline src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java]: I am not sure if I fully understand this. Can you list some examples?

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java [hunk 4]
```diff
@@ -96,7 +102,44 @@ public void checkRoles(RoleChecker checker) throws SecurityException {
...
+                                if (file.isFile() && file.getName().equals("oc")) {
```
Comment (vfreex) [inline src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java]: Does OpenShift Client plugin support Windows slaves? If yes, I think `oc.exe`, `oc.bat`,etc  (see `PATHEXT` environment variables)  should also be checked.

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java [hunk 5]
```diff
@@ -96,7 +102,44 @@ public void checkRoles(RoleChecker checker) throws SecurityException {
...
+                // 2) invoke a shell that then launches the actual 'oc' command, where we update the PATH prior
+                // Our use of the jenkins durable task and bourne/windows scripts did 2), but because of 
+                // https://bugzilla.redhat.com/show_bug.cgi?id=1625518 we analyze the PATH env var and do 1)
+                String path = envVars.get("PATH");
+                // default if running the openshift jenkins images
+                String dirToUse = "/bin/";
+                if (path == null || path.length() == 0) {
+                    LOGGER.warning("PATH not properly set prior to invocation of 'oc'");
+                } else {
+                    String[] dirs = path.split(File.pathSeparator);
```
Comment (vfreex) [inline src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java]: I remember that there is a short circuit: Linux and Windows searches `PATH` only when the specified command doesn't contain `/` (or `\` on Windows).

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java (hunk 4 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java]: Good catch @vfreex - I've never used it, but there is a windows client.  It is `oc.exe` btw.

I'll update.

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java (hunk 5 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java]: Interesting ... in our case it should not matter though.  The command passed to this point will always be `oc` or `oc.exe` ... there will be no fully qualified path to oc.

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java (hunk 2 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/util/ClientCommandRunner.java]: Yeah I like that ... I'll give it a go and assuming it works will change to that form.

File: src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java (hunk 3 above)
Comment (gabemontero) [reply src/main/java/com/openshift/jenkins/plugins/util/ClientCommandBuilder.java]: I'll see about adding some more specific examples in the comment.

And yes, limited use of `QuotedStringTokenizer`, or more machinations here, are the paths I'm considering to deal with the `rsh` / `exec` scenarios vs. the `ASDF=${asdf}` scenario during template parameter passing.

We technically do have a spec, albeit example driven, which just lacked precision wrt these groovy nuances with our API.  I don't see us investing in a formalized spec for this beyond the current help.jelly, example based approach.  Perhaps we start adding warning messages when we detect `ASDF=${asdf}` in the parameter passing that the value should be quoted, and that we'll start strictly enforcing that at a later date, but we'll have to live with whatever short term hack I land on for the time being.  We'll see ...

**Comment by gabemontero** [comment]
changes pushed ... also they crossed the 2 commits enough that I just collapsed it all in one commit

**Comment by gabemontero** [comment]
Saw the oc run failure that just happened in my 3.9 testing, but thought it was due to an old oc I was running .... but it is in fact related to my fix ... updating coming soon I believe.

**Comment by gabemontero** [comment]
ok fully passing for me locally now, including added test case along with existing test cases

**Comment by gabemontero** [comment]
and ok tests passing now

I'll give @bparees and @vfreex a bit to see the latest change before merging 

And I have the updated test case ready once the new version is in the centos image

**Comment by vfreex** [comment]
I am OK with this as long as it fix the current problem we met. Thanks @gabemontero .

**Comment by gabemontero** [comment]
v1.0.18 of the plugin has been initiated with the jenkins update center

**Pull request merged**